tests/t0124_gateway_ipns_record_test.go

```diff
@@ -0,0 +1,113 @@
+		// '
+		{
+			Name: "GET an IPNS record from the gateway",
+			Request: Request().
+				Path("ipns/%s", ipnsKey),
+			Response: Expect().
+				Body(file.RawData()),
+		},
+		// test_expect_success "GET KEY with format=ipns-record and validate key" '
+		// curl "http://127.0.0.1:$GWAY_PORT/ipns/$IPNS_KEY?format=ipns-record" > curl_output_filename &&
+		// ipfs name inspect --verify $IPNS_KEY < curl_output_filename > verify_output &&
+		// test_should_contain "$FILE_CID" verify_output
+		// '
+		// test_expect_success "GET KEY with format=ipns-record has expected HTTP headers" '
+		// curl -sD - "http://127.0.0.1:$GWAY_PORT/ipns/$IPNS_KEY?format=ipns-record" > curl_output_filename 2>&1 &&
+		// test_should_contain "Content-Disposition: attachment;" curl_output_filename &&
+		// test_should_contain "Content-Type: application/vnd.ipfs.ipns-record" curl_output_filename &&
+		// test_should_contain "Cache-Control: public, max-age=3155760000" curl_output_filename
+		// '
+		{
+			Name: "GET KEY with format=ipns-record has expected HTTP headers and valid key",
+			Request: Request().
+				Path("ipns/%s", ipnsKey).
+				Query("format", "ipns-record"),
+			Response: Expect().
+				Headers(
+					Header("Content-Disposition").Contains("attachment;"),
+					Header("Content-Type").Contains("application/vnd.ipfs.ipns-record"),
+					Header("Cache-Control").Contains("public, max-age=3155760000"),
+				).
+				Body(
+					IsIPNSKey(ipnsKey).
+						IsValid().
+						PointsTo("/ipfs/%s", fileCID.String()),
+				),
+		},
+		// test_expect_success "GET KEY with 'Accept: application/vnd.ipfs.ipns-record' and validate key" '
+		// curl -H "Accept: application/vnd.ipfs.ipns-record" "http://127.0.0.1:$GWAY_PORT/ipns/$IPNS_KEY" > curl_output_filename &&
+		// ipfs name inspect --verify $IPNS_KEY < curl_output_filename > verify_output &&
+		// test_should_contain "$FILE_CID" verify_output
+		// '
+		// test_expect_success "GET KEY with 'Accept: application/vnd.ipfs.ipns-record' has expected HTTP headers" '
+		// curl -H "Accept: application/vnd.ipfs.ipns-record" -sD - "http://127.0.0.1:$GWAY_PORT/ipns/$IPNS_KEY" > curl_output_filename 2>&1 &&
+		// test_should_contain "Content-Disposition: attachment;" curl_output_filename &&
+		// test_should_contain "Content-Type: application/vnd.ipfs.ipns-record" curl_output_filename &&
+		// test_should_contain "Cache-Control: public, max-age=3155760000" curl_output_filename
+		// '
+		{
+			Name: "GET KEY with 'Accept: application/vnd.ipfs.ipns-record' has expected HTTP headers and valid key",
+			Request: Request().
+				Path("ipns/%s", ipnsKey).
+				Header("Accept", "application/vnd.ipfs.ipns-record"),
+			Response: Expect().
+				Headers(
+					Header("Content-Disposition").Contains("attachment;"),
+					Header("Content-Type").Contains("application/vnd.ipfs.ipns-record"),
+					Header("Cache-Control").Contains("public, max-age=3155760000"),
+				).
+				Body(
+					IsIPNSKey(ipnsKey).
+						IsValid().
```

Comment:
`IsValid()` is obsolete, right?
```suggestion
					IsIPNSKey(ipnsKey)
```

Reply:
Alternatively, maybe this would be more straightforward
```suggestion
					IPNSKey(ipnsKey).
						IsValid().
```

Reply:
What makes you think it's obsolete?

I would keep it since it's one predicate we'd rather make explicit:

```
x := IsTarfile() // returns a Check function, calling x.Check(bytes) tells you whether a stream of bytes is a tar file,
y := IsTarFile().HasFile("file.txt") // A more precise Check functions

i := IsCarFile()
j := IsCarFile().HasBlock(....id)

m := IsIPNSKey(...)
n := IsIPNSKey(...).IsValid() // A more precise check function that verify the validity of the key
```

Reply:
https://github.com/ipfs/gateway-conformance/blob/16f9bba1698ac4fdab0ee12b4ac06512fe101e46/tooling/ipns/check.go#L17-L23

The validity check is always. The API doesn't allow you to skip it. I guess the questions are: do we ever wanna allow invalid IPNS keys and/or are invalid IPNS keys IPNS keys?

Reply:
@lidel @hacdias do you have an opinion on this?

Reply:
I think we want to be able to test negative scenarios like:
- error on bytes that are not a valid protobuf with ipns record ([spec](https://specs.ipfs.tech/ipns/ipns-record/#record-serialization-format)) (e.g. corrupted mid-flight)
-  error on invalid record  (e.g. valid protobuf, invalid signature)

So I would keep validation in a separate step (`IsValid` func).

btw: See my note on  renaming :upside_down_face:  it -- https://github.com/ipfs/gateway-conformance/pull/37#pullrequestreview-1429480277

```go
IsIPNSRecord(...) // calling x.Check(bytes) tells you whether a stream of bytes is an IPNS key that follows protobuf from https://specs.ipfs.tech/ipns/ipns-record/#record-serialization-format
IsIPNSRecord(...).IsValid() // A more precise check function that verify the validity of the key based on  checks described in https://specs.ipfs.tech/ipns/ipns-record/#record-verification
```

Reply:
@galargh I can see why that makes sense now (we have a `panic("not implemented")` on `false`), but that would create an opportunity for mistakes later: when we implement the `false` case, if we forgot about the default value, then the Check will accept almost any random stream of bytes.

I defaulted to nil, and require `IsValid()` or `IsInvalid()` to be called, let me know what you think.

Reply:
Sounds good :)